### PR TITLE
RavenDB-14284 Don't drop existing connection if we get a new one from…

### PR DESCRIPTION
**Pull request by karmeli87**
… the same source

### CODE REVIEW

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 1]
```diff
@@ -312,17 +317,10 @@ private void AssertValidConnection(IncomingConnectionInfo connectionInfo)
...
-                {
-                    _log.Info(
-                        $"Disconnecting existing connection from {value.FromToString} because we got a new connection from the same source db");
-                }
-
-                IncomingReplicationRemoved?.Invoke(value);
-
-                value.Dispose();
+                throw new InvalidOperationException(
+                    $"connection for this database already exists from {value.ConnectionInfo.SourceUrl}.");
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: Minor: connection -> Connection

**Review by arekpalinski — APPROVED** [review]

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 2]
```diff
@@ -288,6 +291,8 @@ private void ForceTryReconnectAll()
                     }
                 }
             }
+
+            Interlocked.Exchange(ref _reconnectInProgress, 0);
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: Put it inside a finally?

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: I think we want to make this slightly smarter. Do we have any state around the connection?
Can we see when was the last time we got something from it?

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: done

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: We can have the last heartbeat time and last etag from source (and maybe other stuff as well).
Do we want here a better exception? or we want to decide which connection should be dropped?

Comment (ayende) [reply src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: We always want better errors :-)
But yes, if the old connection hasn't had communication recently, we want to drop that and let the new one in

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 3]
```diff
@@ -312,13 +322,15 @@ private void AssertValidConnection(IncomingConnectionInfo connectionInfo)
                     $"Cannot accept the incoming replication connection from {connectionInfo.SourceUrl}, because this node is in passive state.");
             }
 
-            if (_incoming.TryRemove(connectionInfo.SourceDatabaseId, out IncomingReplicationHandler value))
+            if (_incoming.TryGetValue(connectionInfo.SourceDatabaseId, out var value))
             {
+                if (value.LastHeartbeatTicks + TimeSpan.FromMinutes(1).Ticks > Database.Time.GetUtcNow().Ticks)
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: Do we really need to calculate that with the usage of `Ticks`? :) 

- Can we use `new DateTime(value.LastHeartbeatTicks)` ?
- `TimeSpan.FromMinutes(1)` - please move to static readonly field

Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: done

**Pull request merged**